src/base/http/requestparser.cpp

```diff
@@ -30,314 +31,322 @@
+RequestParser::RequestParser()
 {
 }
 
-RequestParser::ErrorCode RequestParser::parseHttpRequest(const QByteArray &data, Request &request)
+RequestParser::ParseResult RequestParser::parse(const QByteArray &data)
 {
-    m_request = Request();
+    // Warning! Header names are converted to lowercase
+    return RequestParser().doParse(data);
+}
 
-    // Parse HTTP request header
+RequestParser::ParseResult RequestParser::doParse(const QByteArray &data)
+{
+    // for simplicity, we don't handle malformed requests which use double `LF` as delimiter
     const int headerEnd = data.indexOf(EOH);
     if (headerEnd < 0) {
         qDebug() << Q_FUNC_INFO << "incomplete request";
-        return IncompleteRequest;
+        return {ParseStatus::Incomplete, Request(), 0};
     }
 
-    if (!parseHttpHeader(data.left(headerEnd))) {
+    const QString httpHeaders = QString::fromLatin1(data.constData(), headerEnd);
+    if (!parseStartLines(httpHeaders)) {
         qWarning() << Q_FUNC_INFO << "header parsing error";
-        return BadRequest;
+        return {ParseStatus::BadRequest, Request(), 0};
     }
 
-    // Parse HTTP request message
-    if (m_request.headers.contains("content-length")) {
-        int contentLength = m_request.headers["content-length"].toInt();
-        if (contentLength < 0) {
-            qWarning() << Q_FUNC_INFO << "bad request: content-length is negative";
-            return BadRequest;
+    const int headerLength = headerEnd + EOH.length();
+
+    // handle supported methods
+    if ((m_request.method == HEADER_REQUEST_METHOD_GET) || (m_request.method == HEADER_REQUEST_METHOD_HEAD))
+        return {ParseStatus::OK, m_request, headerLength};
+    if (m_request.method == HEADER_REQUEST_METHOD_POST) {
+        bool ok = false;
+        const int contentLength = m_request.headers[HEADER_CONTENT_LENGTH].toInt(&ok);
+        if (!ok || (contentLength < 0)) {
+            qWarning() << Q_FUNC_INFO << "bad request: content-length invalid";
+            return {ParseStatus::BadRequest, Request(), 0};
         }
-
-        if (contentLength > static_cast<int>(m_maxContentLength)) {
+        if (contentLength > MAX_REQUEST_SIZE) {
             qWarning() << Q_FUNC_INFO << "bad request: message too long";
-            return BadRequest;
+            return {ParseStatus::BadRequest, Request(), 0};
         }
 
-        QByteArray content = data.mid(headerEnd + EOH.length(), contentLength);
-        if (content.length() < contentLength) {
-            qDebug() << Q_FUNC_INFO << "incomplete request";
-            return IncompleteRequest;
-        }
+        if (contentLength > 0) {
+            const QByteArray httpBodyView = viewMid(data, headerLength, contentLength);
+            if (httpBodyView.length() < contentLength) {
+                qDebug() << Q_FUNC_INFO << "incomplete request";
+                return {ParseStatus::Incomplete, Request(), 0};
+            }
+            if (httpBodyView.length() != contentLength) {
```

Comment:
This condition cannot be true since viewMid returns max contentLength bytes, isn't it?

Reply:
yes, I'll remove this.